File: js/bullet.js
// Projectiles fired by the player and enemies.
class Bullet {
  constructor(x, y, vx, vy, opts = {}) {
    this.x = x;
    this.y = y;
    this.vx = vx;
    this.vy = vy;
    this.radius = opts.radius || 3.5;
    this.damage = opts.damage || 1;
    this.friendly = !!opts.friendly;
    this.color = opts.color || (this.friendly ? "#7df9ff" : "#ff5d5d");
    this.dead = false;
    this.trail = [];
  }

  update(dt, w, h) {
    this.trail.push({ x: this.x, y: this.y });
    if (this.trail.length > 6) this.trail.shift();
    this.x += this.vx * dt;
    this.y += this.vy * dt;
    if (this.x < -20 || this.x > w + 20 || this.y < -20 || this.y > h + 20) this.dead = true;
  }

  draw(ctx) {
    // Glow trail.
    for (let i = 0; i < this.trail.length; i++) {
      const p = this.trail[i];
      ctx.globalAlpha = (i / this.trail.length) * 0.4;
      ctx.fillStyle = this.color;
      ctx.beginPath();
      ctx.arc(p.x, p.y, this.radius * 0.7, 0, Utils.TAU);
      ctx.fill();
    }
    ctx.globalAlpha = 1;

    ctx.shadowColor = this.color;
    ctx.shadowBlur = 12;
    ctx.fillStyle = "#ffffff";
    ctx.beginPath();
    ctx.arc(this.x, this.y, this.radius, 0, Utils.TAU);
    ctx.fill();
    ctx.fillStyle = this.color;
    ctx.beginPath();
    ctx.arc(this.x, this.y, this.radius * 0.6, 0, Utils.TAU);
    ctx.fill();
    ctx.shadowBlur = 0;
  }
}

File: src/bullet.rs
use crate::utils::*;
use macroquad::prelude::*;

/// A projectile fired by the player or an enemy.
pub struct Bullet {
    pub pos: Vec2,
    pub vel: Vec2,
    pub radius: f32,
    pub damage: f32,
    pub from_player: bool,
    pub color: Color,
    pub alive: bool,
    pub trail: f32,
}

impl Bullet {
    pub fn new(pos: Vec2, vel: Vec2, damage: f32, from_player: bool, color: Color) -> Self {
        Self {
            pos,
            vel,
            radius: if from_player { 4.0 } else { 5.0 },
            damage,
            from_player,
            color,
            alive: true,
            trail: 0.0,
        }
    }

    pub fn update(&mut self, dt: f32) {
        self.pos += self.vel * dt;
        self.trail = (self.trail + dt * 8.0).min(1.0);
        if self.pos.x < -20.0
            || self.pos.x > WIDTH + 20.0
            || self.pos.y < -20.0
            || self.pos.y > HEIGHT + 20.0
        {
            self.alive = false;
        }
    }

    pub fn draw(&self) {
        let back = self.pos - self.vel.normalize_or_zero() * self.radius * 3.0;
        draw_line(
            back.x,
            back.y,
            self.pos.x,
            self.pos.y,
            self.radius * 1.2,
            with_alpha(self.color, 0.4),
        );
        draw_glow_circle(self.pos, self.radius, WHITE, self.color);
    }
}
